chitchat/src/state.rs

```diff
@@ -104,6 +119,122 @@ impl NodeState {
             .map(|(key, versioned_value)| (key, versioned_value.value.as_str()))
     }
 
+    pub fn set_max_version(&mut self, max_version: Version) {
+        self.max_version = max_version;
+    }
+
+    // Prepare the node state to receive a delta.
+    // Returns `true` if the delta can be applied. In that case, the node state may be mutated (if a
+    // reset is required) Returns `false` if the delta cannot be applied. In that case, the node
+    // state is not modified.
+    #[must_use]
+    fn prepare_apply_delta(&mut self, node_delta: &NodeDelta) -> bool {
+        if node_delta.from_version_excluded > self.max_version {
+            // This delta is coming from the future.
+            // We probably experienced a reset and this delta is not usable for us anymore.
+            // This is not a bug, it can happen, but we just need to ignore it!
+            info!(
+                node=?node_delta.chitchat_id,
+                from_version=node_delta.from_version_excluded,
+                last_gc_version=node_delta.last_gc_version,
+                current_last_gc_version=self.last_gc_version,
+                "received delta from the future, ignoring it"
+            );
+            return false;
+        }
+
+        if self.max_version > node_delta.last_gc_version {
+            // The GCed tombstone have all been already received.
+            // We won't miss anything by applying the delta!
+            return true;
+        }
+
+        // This delta might be missing tombstones with version within
+        // (`node_state.max_version`..`last_gc_version`].
+        //
+        // It is ok if we don't have the associated values to begin
+        // with.
+        if self.last_gc_version >= node_delta.last_gc_version {
```

Comment:
This will almost never be true given that self.max_version >= self.last_gc_version (so self.last_gc_version >= node_delta.last_gc_version implies self.max_version >= node_delta_last_gc_version so the condition of the previous if will most of the times be true), right?
It seems that it would only be true if self.max_version == self.last_gc_version == node_delta.last_gc_version.

Reply:
It is possible to have max_version < last_gc_version.

If a node gets reset, it will set its last_gc_version to something else, and will receive a partial delta.